File: LeetCode/Normal/1-99/47.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        numLst = []
        length = len(nums)
        visit = [0 for i in range(len(nums))]
        # 进行复制
        tempList = visit[:]

        def dfsSearch(index: int):
            if index == length:
                numLst.append(tempList[:])
            for i in range(length):
                # 去重
                if visit[i] == 1:
                    continue
                # 目的使重复的数字从左到右
                elif i > 0 and nums[i] == nums[i - 1] and visit[i - 1] == 0:
                    continue
                visit[i] = 1
                tempList[index] = nums[i]
                dfsSearch(index + 1)
                visit[i] = 0
        # 注意需要排序 否则不会进行正确的内容从左到右去重
        nums.sort()
        dfsSearch(0)
        return numLst

File: LeetCode/Normal/1-99/47.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    int vis[15] = {0};
    vector<vector<int>> vts;
    vector<int> v;
    void dfs(vector<int> &nums)
    {
        if (v.size() == nums.size())
        {
            vts.push_back(v);
            return;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            //去重
            if (vis[i] || i > 0 && nums[i] == nums[i - 1] && vis[i - 1] == false)
                continue;

            vis[i] = 1;
            v.push_back(nums[i]);
            dfs(nums);
            vis[i] = 0;
            v.pop_back();
        }
    }

    vector<vector<int>> permuteUnique(vector<int> &nums)
    {
        //注意排序才方便去重
        sort(nums.begin(), nums.end());
        dfs(nums);
        return vts;
    }
};

int main()
{
    vector<int> nums = {3, 3, 0, 3};
    Solution s;
    s.permuteUnique(nums);
}